pangeo_forge_recipes/storage.py

```diff
@@ -116,14 +115,15 @@ def size(self, path: str) -> int:
         return self.fs.size(self._full_path(path))
 
     @contextmanager
-    def open(self, path: str, **kwargs) -> Iterator[None]:
+    def open(self, path: str, **kwargs) -> Iterator[fsspec.core.OpenFile]:
```

Comment:
```suggestion
    def open(self, path: str, **kwargs) -> Iterator[io.IOBase]:
```
@martindurant, do I understand [your comment here](https://github.com/pangeo-forge/pangeo-forge-recipes/pull/213/files/c001e710e0a49d2f55476d0a7900819cb26acc2f#r717563338) correctly to mean that the return type of `self.fs.open` is not always an `AbstractBufferedFile`, and therefore this annotation should be changed as proposed here, to be compatible with reading from certain implementations (maybe LocalFileSystem, e.g.)?

According to https://github.com/python/mypy/issues/2984#issuecomment-285721489, I believe mypy should accept subclasses of `BaseClass` for an `Iterator[BaseClass]` annotation.

(If so, we'll need to `import io` in this module, of course.)

Reply:
> the return type of self.fs.open is not always an AbstractBufferedFile

Correct. It is a subclass of AbstractBufferedFile for s3, gcsfs, abfs, ftp, http and others, but can also be LocalFileOpener (io.IOBase), SMBFileOpener (object), SFTPFile (paramiko.BufferedFile), ...

Reply:
However, the following assertion fails, so I do see your point

```python
 with file_opener('/etc/hosts') as fp:
     assert isinstance(fp, fsspec.core.OpenFile)
```

So my question is this: __is there some other type besides `fsspec.core.OpenFile` we can use which captures the fact that this is an `ffspec`-related object__.

I added this test, basically asserting that whatever comes out of `file_opener` has an `fs` attribute:

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c001e710e0a49d2f55476d0a7900819cb26acc2f/tests/test_storage.py#L101

This is required by some of the reference-maker code, e.g. here

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/c001e710e0a49d2f55476d0a7900819cb26acc2f/pangeo_forge_recipes/recipes/reference_hdf_zarr.py#L155

Reply:
Maybe better
```
of = file_opener(..)
assert isinstance(of, fsspec.core.OpenFile)
with of as fp:
   ...
```

Reply:
But that's not how we use `file_opener` in Pangeo Forge. We also invoke it as a context manager.

Reply:
There is no fool-proof method to tell whether the output of the context was created by fsspec. You could check for the few concrete classes that we expect, like AbstractBufferedFile, LocalFileOpener.